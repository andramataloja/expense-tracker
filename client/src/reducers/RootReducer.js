import * as types from "../actions/action-types";

export const initialState = {
  month: new Date().getMonth(),
  year: new Date().getFullYear(),
  doughnutData: [],
  expenses: [],
  index: ""
};

export const rootReducer = (state = initialState, action) => {
  switch (action.type) {
    case types.FETCH_MONTH:
      return {
        ...state,
        month: action.payload
      };
    case types.FETCH_YEAR:
      return {
        ...state,
        year: action.payload
      };
    case types.FETCH_DOUGHNUTDATA:
      return {
        ...state,
        doughnutData: action.payload
      };
    case types.FETCH_EXPENSES:
      return {
        ...state,
        expenses: action.payload
      };
    case types.SET_EXPLODEINDEX:
      return {
        ...state,
        index: action.payload
      };
    default:
      return state;
  }
};
export default rootReducer;
