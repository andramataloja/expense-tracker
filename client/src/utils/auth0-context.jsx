import React, { useState, useEffect, useContext } from "react";
import createAuth0Client from "@auth0/auth0-spa-js";

const DEFAULT_REDIRECT_CALLBACK = () =>
  window.history.replaceState({}, document.title, window.location.pathname);

export const Auth0Context = React.createContext();
export const useAuth0 = () => useContext(Auth0Context);
export const Auth0Provider = ({
  children,
  onRedirectCallback = DEFAULT_REDIRECT_CALLBACK
}) => {
  const [isAuthenticated, setIsAuthenticated] = useState();
  const [user, setUser] = useState();
  const [auth0Client, setAuth0Client] = useState();
  const [isLoading, setIsLoading] = useState(true);

  const config = {
    domain: process.env.REACT_APP_AUTH0_DOMAIN,
    client_id: process.env.REACT_APP_AUTH0_CLIENT_ID,
    redirect_uri: "http://localhost:3000/Dashboard"
  };

  useEffect(() => {
    const initializeAuth0 = async () => {
      const auth0FromHook = await createAuth0Client(config);
      setAuth0Client(auth0FromHook);

      if (
        window.location.search.includes("code=") &&
        window.location.search.includes("state=")
      ) {
        const { appState } = await auth0FromHook.handleRedirectCallback();
        onRedirectCallback(appState);
      }

      const isAuthenticated = await auth0FromHook.isAuthenticated();

      setIsAuthenticated(isAuthenticated);

      if (isAuthenticated) {
        const user = await auth0FromHook.getUser();
        setUser(user);
      }

      setIsLoading(false);
    };
    initializeAuth0();
  }, []);

  const handleRedirectCallback = async () => {
    setIsLoading(true);
    await auth0Client.handleRedirectCallback();
    const user = await auth0Client.getUser();
    setIsLoading(false);
    setIsAuthenticated(true);
    setUser(user);
  };
  return (
    <Auth0Context.Provider
      value={{
        isAuthenticated,
        user,
        isLoading,
        handleRedirectCallback,
        getIdTokenClaims: (...p) => auth0Client.getIdTokenClaims(...p),
        loginWithRedirect: (...p) => auth0Client.loginWithRedirect(...p),
        getTokenSilently: (...p) => auth0Client.getTokenSilently(...p),
        logout: (...p) => auth0Client.logout(...p)
      }}
    >
      {children}
    </Auth0Context.Provider>
  );
};
