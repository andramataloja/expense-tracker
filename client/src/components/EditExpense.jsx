import React, { useEffect, useState } from "react";
import {
  Button,
  TextField,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormControl,
  InputLabel,
  OutlinedInput,
  InputAdornment,
  MenuItem,
  FormGroup,
  Box,
  makeStyles
} from "@material-ui/core";
import "date-fns";
import DateFnsUtils from "@date-io/date-fns";
import {
  MuiPickersUtilsProvider,
  KeyboardDatePicker
} from "@material-ui/pickers";
import { useAuth0 } from "../utils/auth0-context";
import axios from "axios";
import CreateIcon from "@material-ui/icons/Create";
import { useSelector, useDispatch } from "react-redux";
import { fetchExpenses } from "../actions/actions";

const useStyles = makeStyles(theme => ({
  updateButton: {
    color: theme.palette.primary.main
  },
  updateCommands: {
    color: theme.palette.secondary.dark
  }
}));

const EditExpense = props => {
  const classes = useStyles();
  const dispatch = useDispatch();
  const month = useSelector(state => state.month);
  const year = useSelector(state => state.year);
  const [open, setOpen] = useState(false);
  const [values, setValues] = useState({
    amount: props.expense.amount,
    description: props.expense.description
  });
  const [category, setCategory] = useState(props.expense.category_id);
  const [selectedDate, setSelectedDate] = useState(props.expense.date);
  const [categoryList, setCategoryList] = useState([]);

  const handleChange = prop => event => {
    setValues({ ...values, [prop]: event.target.value });
  };

  const formatDate = date => {
    return (
      new Date(date).getFullYear() +
      "-" +
      (new Date(date).getMonth() + 1) +
      "-" +
      new Date(date).getDate()
    );
  };

  const handleSubmit = event => {
    event.preventDefault();
    //const user_id = user.user_id
    console.log("enne:", selectedDate);
    const data = {
      description: values.description,
      amount: values.amount,
      date: formatDate(selectedDate),
      //      user_id: user_id,
      category_id: category,
      expense_id: props.expense.expense_id
    };
    console.log("data to submit:", data);
    axios
      .put("/put/expense", data)
      .then(() => {
        setOpen(false);
        axios
          .get("/allexpensesbydate", {
            params: { month: month + 1, year: year }
          })
          .then(res => {
            res.data.length !== 0
              ? dispatch(fetchExpenses(res.data))
              : dispatch(fetchExpenses([]));
          })
          .catch(err => console.log(err));
      })
      .catch(err => console.log(err));
  };

  useEffect(() => {
    axios
      .get("/get/categories")
      .then(res => (res.data.length !== 0 ? setCategoryList(res.data) : null))
      .catch(err => console.log(err));
  }, []);

  return (
    <Box>
      <CreateIcon
        className={classes.updateButton}
        onClick={() => setOpen(true)}
      />
      <Dialog
        open={open}
        onClose={() => setOpen(false)}
        aria-labelledby="form-dialog-title"
      >
        <DialogTitle id="form-dialog-title">Edit Expense</DialogTitle>
        <DialogContent>
          <FormGroup>
            <FormControl variant="outlined" margin="dense">
              <InputLabel htmlFor="component-outlined">Description</InputLabel>
              <OutlinedInput
                id="component-outlined"
                value={values.description}
                onChange={handleChange("description")}
                label="Description"
              />
            </FormControl>
            <FormControl variant="outlined" margin="dense">
              <InputLabel htmlFor="outlined-adornment-amount">
                Amount
              </InputLabel>
              <OutlinedInput
                id="outlined-adornment-amount"
                value={values.amount}
                onChange={handleChange("amount")}
                startAdornment={
                  <InputAdornment position="start">â‚¬</InputAdornment>
                }
                label="Amount"
                type="number"
                inputProps={{ min: "0" }}
              />
            </FormControl>
            <MuiPickersUtilsProvider utils={DateFnsUtils}>
              <KeyboardDatePicker
                autoOk
                disableToolbar
                variant="inline"
                inputVariant="outlined"
                format="dd/MM/yyyy"
                margin="dense"
                id="date-picker-inline"
                label="Date"
                value={selectedDate}
                onChange={date => setSelectedDate(date)}
                KeyboardButtonProps={{
                  "aria-label": "change date"
                }}
              />
            </MuiPickersUtilsProvider>
            <TextField
              id="outlined-select-category"
              select
              label="Category"
              value={category}
              onChange={event => setCategory(event.target.value)}
              variant="outlined"
              margin="dense"
            >
              {categoryList.map(option => (
                <MenuItem key={option.category_id} value={option.category_id}>
                  {option.category_name}
                </MenuItem>
              ))}
            </TextField>
          </FormGroup>
        </DialogContent>
        <DialogActions>
          <Button
            onClick={() => setOpen(false)}
            className={classes.updateCommands}
          >
            Cancel
          </Button>
          <Button onClick={handleSubmit} className={classes.updateCommands}>
            Update
          </Button>
        </DialogActions>
      </Dialog>
    </Box>
  );
};
export default EditExpense;
