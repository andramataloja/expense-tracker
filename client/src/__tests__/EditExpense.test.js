import React from "react";
import {
  cleanup,
  render,
  fireEvent,
  screen,
  wait
} from "@testing-library/react";
import EditExpense from "../components/EditExpense";
import { useAuth0 } from "../utils/auth0-context";
import { Provider } from "react-redux";
import { act } from "react-dom/test-utils";
import { initialState } from "../reducers/RootReducer";
import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import mockAxios from "axios";

const API = "http://localhost:3000";

const user = {
  email: "johndoe@me.com",
  email_verified: true,
  sub: "google-oauth2|12345678901234"
};

jest.mock("../utils/auth0-context");
jest.mock("../store.js");

describe("Edit Expense", () => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);

  afterEach(cleanup);
  beforeEach(() => {
    useAuth0.mockReturnValue({
      isAuthenticated: true,
      user,
      logout: jest.fn(),
      loginWithRedirect: jest.fn()
    });
    require("mutationobserver-shim");
    mockAxios.get.mockImplementationOnce(() => Promise.resolve(mockData));
  });

  const mockData = {
    data: [
      {
        category_id: 3,
        category_name: "Entertainment",
        icon: "Entertainment",
        fill: "#EA6E6E"
      },
      {
        category_id: 1,
        category_name: "Food",
        icon: "Food",
        fill: "#FCD246"
      }
    ]
  };

  const expense = {
    amount: 8,
    description: "Snacks",
    category_id: 1,
    date: "17/03/2020"
  };

  it("fetches successfully data from an API", async () => {
    expect(mockAxios.get(`${API}/get/categories`)).resolves.toEqual(mockData);
    expect(mockAxios.get).toHaveBeenCalledWith(`${API}/get/categories`);
  });

  it("should take a snapshot of edit button", async () => {
    const store = mockStore();
    const { asFragment } = render(
      <Provider store={store}>
        <EditExpense expense={expense} />
      </Provider>
    );
    await act(wait);

    expect(asFragment(<EditExpense />)).toMatchSnapshot();
  });

  it("edit dialog renders", async () => {
    const store = mockStore();
    render(
      <Provider store={store}>
        <EditExpense expense={expense} />
      </Provider>
    );
    await act(wait);

    fireEvent.click(screen.getByTestId("edit-button"));
    expect(screen.getByText("Edit Expense")).toBeVisible();
    expect(screen.getByText("Cancel")).toBeVisible();
    expect(screen.getByText("Update")).toBeVisible();
  });

  it("edit dialog opens and closes", async () => {
    const store = mockStore();
    render(
      <Provider store={store}>
        <EditExpense expense={expense} />
      </Provider>
    );
    await act(wait);

    expect(screen.queryByTestId("edit-dialog")).not.toBeInTheDocument();
    fireEvent.click(screen.getByTestId("edit-button"));
    expect(screen.getByTestId("edit-dialog")).toBeVisible();
    fireEvent.click(screen.getByText("Cancel"));
    await wait(() =>
      expect(screen.queryByTestId("edit-dialog")).not.toBeInTheDocument()
    );
  });

  it("Edit fields change", async () => {
    const store = mockStore();
    const handleChange = jest.fn();
    render(
      <Provider store={store}>
        <EditExpense onChange={handleChange} expense={expense} />
      </Provider>
    );
    await act(wait);

    fireEvent.click(screen.getByTestId("edit-button"));
    const description = screen.getByLabelText("Description");
    fireEvent.change(screen.getByLabelText("Description"), {
      target: { value: "TEST VALUE" }
    });
    expect(description).toHaveValue("TEST VALUE");

    const amount = screen.getByLabelText("Amount");
    fireEvent.change(amount, {
      target: { value: 5 }
    });
    expect(amount).toHaveValue(5);

    /*  const date = screen.getByLabelText("Date");
    console.log(screen.debug(date));
    fireEvent.change(date, {
      target: { value: "05/02/2020" }
    });
    expect(date).toHaveValue("05/02/2020"); */

    const category = screen.getByTestId("category").querySelector("input");
    fireEvent.change(category, {
      target: { value: "3" }
    });
    expect(category).toHaveValue("3");
  });

  it("Handle submit", async () => {
    /*  const store = mockStore();
    const handleSubmit = jest.fn();
    render(
      <Provider store={store}>
        <EditExpense onSubmit={handleSubmit} />
      </Provider>
    );
    await act(wait);
    fireEvent.click(screen.getByTestId("edit-button"));
    fireEvent.click(screen.getByText("Update")); */
  });
});
