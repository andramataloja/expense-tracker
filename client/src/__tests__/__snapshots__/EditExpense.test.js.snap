// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Expense should take a snapshot of edit button 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-5"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root makeStyles-updateButton-1"
      data-testid="edit-button"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
